### ISSUE TITLE
improve time performance

### ISSUE THREAD

**Issue by winterland1989** [haskell/time#54]
I'm quite pleased by `time` 's portability(comparing to `unix-time`) , but the performance can be improved IMHO:

+ We shouldn't create thunks during conversions, i propose mark the field of `TimeOfDay`, `LocalTime`, `UTCTime` with strict pragmas, some field also should be unpacked. 

+ We should add `POSIXTimeRaw` like the one from `unix-time`, with a `CTime` and `Int32` field, it's faster, when year 2038 come we have to deal with this anyway.

+ I'd like to add `Text` based formatter and parser, but that can be done in a seperate package if `text` is too heavy to be depended on.

**Comment by hvr** [comment]
Btw the `text` dependency is rather undesirable at this point since it would mean that GHC needs to bundle `text` as a boot-lib dependency (which we don't yet). (fwiw, I'm sympathetic to optimising `time`, especially if there's low-hanging fruits, as those will instantly benefit large parts of the ecosystem given `time` is used by almost every package requiring facilities related to time handling)

**Comment by winterland1989** [comment]
I agree with that `Text` based formatter/parser should be done seperately. But the first two point can be done without breakage, is there any objection on those proposal?

**Comment by AshleyYakeley** [comment]
Could you clarify your intention for `POSIXTimeRaw`?

**Comment by winterland1989** [comment]
Hi Ashley! The intention is to improve speed on common timestamp usage. Image if we want get a formatted timestamp like "Mon, 28 Nov 2016 18:05:07 +0800" using `time`, we can use `getZonedTime`, but it's quite slow because:

+ `getZonedTime` first calls `getCurrentTime` to obtain `UTCTime`, which in turn calls `getPOSIXTime`, `getPOSIXTime` have to convert POSIX presentation(CTime and Int32) to `Integer`.

+ `getZonedTime` then calls `getTimeZone` which convert this `Integer` back to POSIX presentation and pass it to FFI.

If we have `POSIXTimeRaw`, we can save these conversions by provide `getPOSIXTimeRaw` and define `getZonedTime` with it. In fact `POSIXTime` shouldn't be defined with `NominalDiffTime` at the first place because it's supposed reflect POSIX standards. But i think changing it will cause too much breakage, so `POSIXTimeRaw` is introduced.

**Comment by winterland1989** [comment]
Basically i'd like to take `POSIXTime` as a glue type between high level type such as `UTCTime/NominalDiffTime` and POSIX raw time presentation, thus we shouldn't hide `CTime/Int32` in `POSIXTime`.

We can get an `UTCTime` directly from `POSIXTimeRaw` thus saving another operation on `Integer`.

**Comment by winterland1989** [comment]
I have done a quick experiment in #55 , we don't have to add new types if
 that's not desired. the `getCurrentTime` is improved almost 15X comparing old implementations.

**Comment by AshleyYakeley** [comment]
Can we replace both `POSIXTime` and `POSIXTimeRaw` with a newtype of `Int64` of nanoseconds?

**Comment by AshleyYakeley** [comment]
I think that's +/- 25m years.

**Comment by AshleyYakeley** [comment]
Ideally the `Fixed` type would be generalised on its underlying type.

**Comment by winterland1989** [comment]
Single `Int64` is not enough i'm afraid, since nanosecond is too small :

```
> (maxBound :: Int64) `div` (1000000000 * 60 * 60 * 12 * 365)
> 584
```

It's up to you to decide if we want to keep `POSIXTime`, personally i want to keep it for two reason:

+ Backward compatibility.
+ High resolution (`ps`), even it's not possible to fetch a time value with such high resolution.

In practice i assume most user will use `POSIXTimeRaw` once my patch is merged, since it's very fast and accurate enough.

**Comment by AshleyYakeley** [comment]
In principle, I dislike having two types which have the same meaning. The intent of `POSIXTime` is only to represent time as obtained from a system clock that cannot represent leap-seconds: it's already supposed to be "raw". For this reason, I would prefer to use your faster `POSIXTimeRaw` implementation for `POSIXTime`, rather than just being a synonym of `NominalDiffTime`. You already provide `Eq` and `Ord` instances. I would add a `Show` instance, and I think those are the only ones necessary.

Sadly `Int128` is not available as a scalar type, or I would recommend that as the implementation.

**Comment by winterland1989** [comment]
That's even better! I can't agreed with you more that `POSIXTime` should reflect POSIX raw time presentation. I'll remove old `POSIXTime` and change `POSIXTimeRaw` to `POSIXTime` soon.

**Comment by winterland1989** [comment]
Done in 45378ef , I also removed the benchmark cabal stanze, if anyone got any idea on how to benchmark time properly, please tell me.

**Comment by hvr** [comment]
As I'm a bit too lazy to do the math: what's going to be the new effective range of `POSIXTime` switching to Int64?

**Comment by AshleyYakeley** [comment]
So this is what's now in the PR:

    data POSIXTime = POSIXTime
        { ptSeconds ::     {-# UNPACK #-} !Int64
        , ptNanoSeconds :: {-# UNPACK #-} !Int64
        }

`ptNanoSeconds` could fit in `Word32`, but it probably doesn't matter.

`ptSeconds` must be signed, to represent time before the epoch (beginning of 1970 UT)

Range is about +/- 292 billion years (I think...).

**Comment by winterland1989** [comment]
Yes, and since FFI type `CTime` and `CLong` wrap `Int64` inside, we just reuse `Int64`.

**Comment by AshleyYakeley** [comment]
Is this guaranteed on all systems?

**Comment by winterland1989** [comment]
I think so, since we have fixed 64bit width, but underline c call(`gettimeofday`) can't guaranteed always return 64bit value if i didn't get wrong, in another word, we can't fix 2038 problem if the underline system have this problem.

**Comment by AshleyYakeley** [comment]
I don't believe this is guaranteed. `CTime` corresponds to `time_t`, which may be 32 bit (in which case the system has the 2038 problem). http://stackoverflow.com/questions/471248/what-is-ultimately-a-time-t-typedef-to

Maybe we should do this?

    data POSIXTime = POSIXTime
        { ptSeconds ::     {-# UNPACK #-} !CTime
        , ptNanoSeconds :: {-# UNPACK #-} !CLong
        }

**Comment by winterland1989** [comment]
But this doesn't improve anything to 2038 problem IMHO, the cast is done after we get `CTime`, it's no difference with `Int64` from haskell view, only a extra new type.

**Comment by hvr** [comment]
@AshleyYakeley I'm not sure it's a good idea to make POSIXTime's range platform dependent; e.g. if you get a POSIX `.tar` with entries having a  POSIXTime , you'd want to get the same values representable regardless of which platform you compiled it on... IMO :-)

**Comment by AshleyYakeley** [comment]
OK, `POSIXTime` should be platform-independent. On a machine where `time_t` is 32 bit, `getPOSIXTime` etc. will be wrong after 2038, and that's unavoidable.

**Comment by winterland1989** [comment]
Since now we made a breaking change, I want to have a discussion on changing `Day` representation, I argue that `Int64` can hold day number more than human may survive, it's reasonable to change `Day` 's representation to `Int64`.

**Comment by AshleyYakeley** [comment]
That's a separate issue, with potentially a much larger impact.

**Comment by winterland1989** [comment]
It indeed has a larger impact, but since this issue concerns about performance, if we don't discuss it now, it's likely that it will never be discussed afterwards.

**Comment by AshleyYakeley** [comment]
I think a PR involving a change to the `Day` type is going to require much wider consultation.

**Comment by winterland1989** [comment]
Yes, is there any workflow I should use other than github for this?

**Comment by AshleyYakeley** [comment]
Probably not, but it should be a separate issue / PR.

**Comment by winterland1989** [comment]
OK, that make sense, another thing i want to do is experiment new `FormatTime` formula, the original 

```haskell
type NumericPadOption = Maybe Char

class FormatTime t where
    formatCharacter :: Char -> Maybe (TimeLocale -> Maybe NumericPadOption -> t -> String) 
```

is quite complex and force `formatTime` to use `++` in a rather inefficient way. I'd like to borrow `thyme` 's formula:

```haskell
class FormatTime t where
    showsTime :: TimeLocale -> t -> (Char -> ShowS) -> Char -> ShowS
```

If benchmarks could be my favor, would you want a separate PR or just in #55 ?

**Comment by AshleyYakeley** [comment]
If there's anything else to be done here, please open a new issue or PR.
